test/core/connection.test.ts

```diff
@@ -241,21 +280,80 @@ describe('connection', () => {
+        const stateChanges: ConnectionStatusChange[] = [];
         connection.onStatusChange((status) => {
-          stateChanges.push(status.current);
+          stateChanges.push(status);
         });
 
         // Transition to a disconnected state
-        context.emulateStateChange({ current: 'disconnected', previous: 'connected' });
+        const disconnectError = new Ably.ErrorInfo('error', 500, 99999);
+        context.emulateStateChange({ current: 'disconnected', previous: 'connected', reason: disconnectError });
 
         // Wait for longer than the transient timeout
         void new Promise((resolve) => setTimeout(resolve, 6000)).then(() => {
           // Transition back to a connected state
           context.emulateStateChange({ current: 'connected', previous: 'disconnected' });
 
           // Assert that we have only seen the connected state
-          expect(stateChanges).toEqual([ConnectionStatus.Disconnected, ConnectionStatus.Connected]);
+          expect(stateChanges.map((change) => change.current)).toEqual([
+            ConnectionStatus.Disconnected,
+            ConnectionStatus.Connected,
+          ]);
+
+          // The first change should have the disconnect error
+          expect(stateChanges[0]?.error).toEqual(disconnectError);
+
+          done();
+        });
+      }),
+    10000,
+  );
+
+  it<TestContext>(
+    'emits disconnections after a period of time to an intermediate state',
+    (context) =>
+      new Promise<void>((done) => {
+        // Start the channel in a connected state
+        vi.spyOn(context.realtime.connection, 'state', 'get').mockReturnValue('connected');
+        vi.spyOn(context.realtime.connection, 'errorReason', 'get').mockReturnValue(
+          new Ably.ErrorInfo('error', 500, 99999),
+        );
+
+        const connection = new DefaultConnection(context.realtime, makeTestLogger());
+        expect(connection.status).toEqual(ConnectionStatus.Connected);
+
+        // Set a listener that stores the state change
+        const stateChanges: ConnectionStatusChange[] = [];
+        connection.onStatusChange((status) => {
+          stateChanges.push(status);
+        });
+
+        // Transition to a disconnected state
+        const disconnectError = new Ably.ErrorInfo('error', 500, 99999);
+        context.emulateStateChange({ current: 'disconnected', previous: 'connected', reason: disconnectError });
+
+        // Now transition to a connecting state
+        context.emulateStateChange({ current: 'connecting', previous: 'disconnected' });
+
+        // Now transition to a disconnected state
+        context.emulateStateChange({ current: 'disconnected', previous: 'connecting' });
+
+        // Now transition to a connecting state
+        context.emulateStateChange({ current: 'connecting', previous: 'disconnected' });
+
+        // Wait for longer than the transient timeout
+        void new Promise((resolve) => setTimeout(resolve, 8000)).then(() => {
```

Comment:
same applies here regarding async fn and use of `void` operator

Reply:
Done